### PR TITLE
1 domain adaptation

### LINKED ISSUE #1: fine tune model for domain adaptation

### PR DESCRIPTION
**Pull request by india-kerle**
---

# Description

This PR adds a few utils scripts and the flow to train a transformer model for domain adaptation.

Fixes # (issue)

This PR closes #1 

# Instructions for Reviewer

There's a fair amount of set up to getting your environment ready but i've added it all in the [README.md under Setup](https://github.com/nestauk/ojd_daps_language_models/tree/1_domain_adaptation#setup). One annoying thing I also had to do to use batch is [follow these instructions also](https://github.com/nestauk/ojd_daps_language_models/tree/1_domain_adaptation/ojd_daps_language_models/pipeline#sparkles-setting-up-with-aws-and-metaflow) to deal with the dot in my laptop username. You can see if you need to do that by running `whoami` in your terminal.

Finally, I've also written the readmes like we will add more fine tuned models but honestly that's a hard tbd! 

In order to run the flow not in production, hash out the batch decorator and run:

`python ojd_daps_language_models/pipeline/train_model/ojobert_flow.py run`

If you want to test the GPU, you can make the training size smaller by changing the `bert_train_size` in the `training.yaml` config file. 

In order to test functions used in the flow:

`pytest ojd_daps_language_models/pipeline/train_model/tests/test_ojobert_flow.py`  

Please pay special attention to ...

I've run the script locally and I was also able to run it using a GPU but it wouldn't move on from the `task is starting (status RUNNABLE)... ` stage as I didn't see if there was available resource. 

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,12 +1,39 @@
...
+This repo contains the metaflows used to fine-tune transformer models with job adverts for a variety of downstream tasks, including:
+
+1. next-sentence prediction
+2. masked language modelling
+3. skill semantic similarity
+4. named entity recognition
+
+## :cupid: Using fine-tuned model checkpoints
+
+The fine-tuned models (and their associated model cards) can be accessed via huggingface's hub.
```
Comment (india-kerle) [inline README.md:20]: I haven't added much here as we don't have models on huggingface! I'm just feeling aspirational...

File: ojd_daps_language_models/pipeline/train_model/ojobert_flow.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+        # load data
+        sent_files = all_sent_files[:sent_files_num]
+
+        sents = []
+        for sent_file in sent_files:
+            file = load_s3_data(BUCKET_NAME, sent_file)
+            sents.extend(list(chain(*file)))
+
+        job_sents = list(set(sents))[1:]  # remove empty string
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/ojobert_flow.py]: so is it always true empty strings will come first in a set? I didn't know this. Could also remove any empty strings in the sentence splitting script, which may also free up a little memory from the output files too?

File: ojd_daps_language_models/pipeline/train_model/ojobert_flow.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+        # convert into huggingface Dataset to use multi-threading
+        self.job_sents_dataset = Dataset.from_pandas(pd.DataFrame({"text": job_sents}))
+
+        logger.info(f"Loaded {len(job_sents)} job sentences from S3.")
+
+        self.next(self.prepare_training_data)
+
+    # make sure we have at least 100k job adverts in the training data
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/ojobert_flow.py]: should this comment be deleted?

File: README.md [hunk 4]
```diff
@@ -1,12 +1,39 @@
...
 - Meet the data science cookiecutter [requirements](http://nestauk.github.io/ds-cookiecutter/quickstart), in brief:
   - Install: `direnv` and `conda`
 - Run `make install` to configure the development environment:
   - Setup the conda environment
   - Configure `pre-commit`
+- Download spacy model: `python -m spacy download en_core_web_sm`
+- install Pytorch: `conda install pytorch torchvision -c pytorch` (if you are using mac OS x 13.4 operating system - `pip install torch`)
+- Set up batch processing with Metaflow
+- Sign into huggingface hub to push models to huggingface
+
```
Comment (lizgzil) [inline README.md:36]: I also needed to run
```
export LC_ALL="en_GB.UTF-8"
```

(I was getting an error `Fatal Python error: config_get_locale_encoding: failed to get the locale encoding: nl_langinfo(CODESET) failed` and found the fix in [this](https://stackoverflow.com/questions/60557160/python3-8-fails-with-fatal-python-error-config-get-locale-encoding)

File: ojd_daps_language_models/pipeline/train_model/ojobert_flow.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    seed = Parameter("random_seed", default=42)
+
+    @step
+    def start(self):
+        """
+        Starts the flow and sets parameters based on whether flow is in production or not.
+        """
+        self.batch_size = 64
+        if self.production:
+            self.train_size = 100000
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/ojobert_flow.py]: should this be `CONFIG["bert_train_size"]`?

File: ojd_daps_language_models/pipeline/train_model/ojobert_flow.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    mlm_probability = Parameter("mlm_probability", default=CONFIG["mlm_probability"])
+    sents_per_file = Parameter("sents_per_file", default=CONFIG["sents_per_file"])
+    seed = Parameter("random_seed", default=42)
+
+    @step
+    def start(self):
+        """
+        Starts the flow and sets parameters based on whether flow is in production or not.
+        """
+        self.batch_size = 64
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/ojobert_flow.py]: should this be `CONFIG["bert_batch_size"]`

File: ojd_daps_language_models/pipeline/train_model/ojobert_flow.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        random.seed(self.seed)
+        random.shuffle(all_sent_files)
+
+        # load data
+        sent_files = all_sent_files[:sent_files_num]
+
+        sents = []
+        for sent_file in sent_files:
+            file = load_s3_data(BUCKET_NAME, sent_file)
+            sents.extend(list(chain(*file)))
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/ojobert_flow.py]: something handy for testing quickly might be to add after this:
```
if self.train_size < self.sents_per_file:
   sents = sents[:self.train_size]
```

Comment (lizgzil) [reply ojd_daps_language_models/pipeline/train_model/ojobert_flow.py]: I had to set train_size=100 though as with 10 it complained there was not enough test data

File: ojd_daps_language_models/pipeline/train_model/ojobert_flow.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        save_to_s3(
+            BUCKET_NAME,
+            eval_results,
+            os.path.join(
+                model_eval_path, f"{output_dir}_{date}_evaluation_results.json"
+            ),
+        )
+
+        # save model locally
+        trainer.save_model(output_dir)
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/ojobert_flow.py]: could create a little bash script to then move this whole folder of files to S3

File: README.md (hunk 4 above)
Comment (lizgzil) [reply README.md:36]: Also for batch:

I fixed the `Could not find the location of the datastore -- did you correctly set the METAFLOW_DATASTORE_SYSROOT_S3 environment variable?` error with:

I opened `.envrc` and added the line
```
export METAFLOW_PROFILE=ds-cookiecutter
```

then did 
```
direnv allow .
eval "$(direnv hook zsh)"
direnv reload
eval "$(direnv hook zsh)"
```
(not sure what order to do these in though)

File: ojd_daps_language_models/pipeline/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+There are a number of flows in the `train_model` directory that fine-tune large language models for a number of different tasks.
+
+### :diamond_shape_with_a_dot_inside: Fine-tuning for domain language adaptation
+
+This script fine-tunes a DistilBERT model on job advertisements for domain adaptation for the purposes of next sentence prediction and masked language modelling.
+
+To run the flow (with default model parameters, as defined in `config/training.yaml`):
+
+`python ojd_daps_language_models/pipeline/train_model/ojobert_flow.py run`
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/README.md]: production defaults to False
`python ojd_daps_language_models/pipeline/train_model/ojobert_flow.py run --production True`

**Review by lizgzil — APPROVED** [review]
omg this is really excellent! I hadn't been keeping up with this repo - so it's great to see how much youve done.

I successfully ran:
```
python ojd_daps_language_models/pipeline/train_model/ojobert_flow.py run
```
(but with train_size=100 and no batch, I also added my suggestion of the `sents = sents[:train_size]` thing to speed it up).

It all ran and outputted the saved model, but there was a weird hiccup right at the end:
```
2023-06-09 17:09:11.872 [1686326892567481/end/6 (pid 40061)] Task finished successfully.
2023-06-09 17:09:11.873 Done!
Traceback (most recent call last):
  File "/Users/elizabethgallagher/miniconda3/envs/ojd_daps_language_models/lib/python3.9/site-packages/metaflow/_vendor/click/core.py", line 792, in main
    ctx.exit()
  File "/Users/elizabethgallagher/miniconda3/envs/ojd_daps_language_models/lib/python3.9/site-packages/metaflow/_vendor/click/core.py", line 558, in exit
    raise Exit(code)
metaflow._vendor.click.exceptions.Exit: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/elizabethgallagher/Code/ojd_daps_language_models/ojd_daps_language_models/pipeline/train_model/ojobert_flow.py", line 225, in <module>
    OjoBertFlow()
  File "/Users/elizabethgallagher/miniconda3/envs/ojd_daps_language_models/lib/python3.9/site-packages/metaflow/flowspec.py", line 108, in __init__
    cli.main(self)
  File "/Users/elizabethgallagher/miniconda3/envs/ojd_daps_language_models/lib/python3.9/site-packages/metaflow/cli.py", line 1172, in main
    start(auto_envvar_prefix="METAFLOW", obj=state)
  File "/Users/elizabethgallagher/miniconda3/envs/ojd_daps_language_models/lib/python3.9/site-packages/metaflow/_vendor/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/elizabethgallagher/miniconda3/envs/ojd_daps_language_models/lib/python3.9/site-packages/metaflow/_vendor/click/core.py", line 810, in main
    sys.exit(e.exit_code)
SystemExit: 0
```
🤷‍♀️  Google says something to do with a dependency issue?

## Batch

I ran
```
python ojd_daps_language_models/pipeline/train_model/ojobert_flow.py --datastore=s3 run
```
with the batch decorator included.

and got:
```
  raise ServiceException(
metaflow.plugins.metadata.service.ServiceException: Metadata request (/flows/OjoBertFlow) failed (code 500): {"message": "Internal server error"}
```
It's probably because I haven't set something up properly? So didn't even get close to the unmoving "task is starting (status RUNNABLE)..." step.

just comparing the last time I used batch I had:
```
@batch(
        queue="job-queue-GPU-nesta-metaflow",
        image="metaflow-pytorch",
        memory=60000,
        cpu=8,
    )
```
so might be worth adding the `image` arg to see if that sorts the neverending "task is starting" step. Not sure why it would though!

Anyway - I had minor comments which might be worth addressing, but it looks good and it must be a metaflow issue rather than something here.

Don't know if you want to wait for it to work on batch before merging - but I'll leave that to you

**Pull request merged**